src/library_syscall.js

```diff
@@ -752,6 +752,9 @@ var SyscallsLibrary = {
         if (arg < 0) {
           return -{{{ cDefs.EINVAL }}};
         }
+        while (FS.streams[arg]) {
+          arg++;
+        }
```

Comment:
Ah I see, this is very different to the behavior of the `dup` syscall then.

Reply:
Yes, it's a bit upsetting that there are four different ways to `dup` file descriptors:

1. `dup` -- simple, just give me any file descriptr
2. `dup2` -- use a specific file descriptor, overwriting any existing stream at that descriptor, useful for redirecting standard streams 
3. `dup3` -- `dup2` plus you can at the same time force `CLOEXEC` to be set, saving an extra system call. Setting `CLOEXEC` is a good standard practice. But it's a bit weird though because it has to force a specific descriptor but you probably don't want to set `CLOEXEC` on the standard streams??
4. `fcntl` -- you can specify a least possible descriptor and it will give you an unused one above it. Also, you can set `O_CLOEXEC` like `dup3`. Most often used when setting `CLOEXEC` with minimum file descriptor set to 0 because people don't care which file descriptor they get but they want `CLOEXEC`.

So of the options, `dup3` seems kind of useless and `fcntl` seems like a fixed replacement for it. And I guess at some point they decided to stop using up short names for file descriptor operations and combine them all into one mega function, hence `fcntl`? Of course `CLOEXEC` does nothing at all on emscripten I don't think so we just need to make sure we don't break anything.